### PR TITLE
types: fix the bug about the wrong query result for decimal type 

### LINKED ISSUE #22496: The wrong query result for decimal type
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

# create a test table in TiDB
MySQL [test]> create table t12(d decimal(15,2));
MySQL [test]> insert into t12 values('1,999.00');
MySQL [test]> select * from t12;
+------+
| d    |
+------+
| 1.00 |
+------+
1 row in set (0.00 sec)

### 2. What did you expect to see? (Required)
show the error.

### 3. What did you see instead (Required)
select * from t12; ==> 1.00

### 4. What is your TiDB version? (Required)
5.7.25-TiDB-v4.0.9

### PR DESCRIPTION
**Pull request by jyz0309**
### What problem does this PR solve?

Issue Number: close #22496 

Problem Summary:

MySQL [test]> create table t12(d decimal(15,2));
MySQL [test]> insert into t12 values('1,999.00');
MySQL [test]> select * from t12;
+------+
| d |
+------+
| 1.00 |
+------+
1 row in set (0.00 sec)

The select about decimal type show the wrong result.
### What is changed and how it works?
What's Changed:
update the `FromString` function for MyDecimal for the situation like `"1,999.00"` and `"1,999,999.00"`

### Related changes
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)

### Release note 
- types: fix the bug about the wrong query result for decimal type

### CODE REVIEW

**Comment by jyz0309** [comment]
@qw4990 PTAL

**Comment by lysu** [comment]
image

https://www.db-fiddle.com/

question: does mysql support comma number? @jyz0309

**Comment by jyz0309** [comment]
> https://www.db-fiddle.com/

oh, I didn't notice this, my bad-.-

**Comment by jyz0309** [comment]
@lysu @XuHuaiyu PTAL

File: types/mydecimal.go [hunk 1]
```diff
@@ -407,7 +407,7 @@ func (d *MyDecimal) FromString(str []byte) error {
 		str = str[1:]
 	}
 	var strIdx int
-	for strIdx < len(str) && isDigit(str[strIdx]) {
+	for strIdx < len(str) && (isDigit(str[strIdx])) {
```
Comment (lysu) [inline types/mydecimal.go:410]: maybe can avoid this modification

File: types/mydecimal.go [hunk 2]
```diff
@@ -443,8 +446,8 @@ func (d *MyDecimal) FromString(str []byte) error {
 	var word int32
 	var innerIdx int
 	for digitsInt > 0 {
-		digitsInt--
 		strIdx--
+		digitsInt--
```
Comment (lysu) [inline types/mydecimal.go:450]: ditto

File: types/convert_test.go [hunk 3]
```diff
@@ -243,6 +243,12 @@ func (s *testTypeConvertSuite) TestConvertType(c *C) {
 	v, err = Convert("-10000", ft)
 	c.Assert(terror.ErrorEqual(err, ErrOverflow), IsTrue, Commentf("err %v", err))
 	c.Assert(v.(*MyDecimal).String(), Equals, "-9999.9999")
+	v, err = Convert("1,999.00", ft)
+	c.Assert(terror.ErrorEqual(err, ErrBadNumber), IsTrue, Commentf("err %v", err))
+	c.Assert(v.(*MyDecimal).String(), Equals, "0.0000")
+	v, err = Convert("1,999,999.00", ft)
+	c.Assert(terror.ErrorEqual(err, ErrBadNumber), IsTrue, Commentf("err %v", err))
+	c.Assert(v.(*MyDecimal).String(), Equals, "0.0000")
```
Comment (lysu) [inline types/convert_test.go:251]: how about add a test case to cover 

```
set sql_mode='';
insert into t12 values('1,999.00');
```

result~?

Comment (jyz0309) [reply types/convert_test.go:251]: ok, can I add this test case in `/executor/write_test.go`?

Comment (jyz0309) [reply types/convert_test.go:251]: > how about add a test case to cover
> 
> ```
> set sql_mode='';
> insert into t12 values('1,999.00');
> ```
> 
> result~?

ok, can I add this test case in `/executor/write_test.go`?

Comment (lysu) [reply types/convert_test.go:251]: emm... yes~ maybe it's incompatible with mysql when in sql_mode=''..

Comment (lysu) [reply types/convert_test.go:251]: maybe we can ignore `sql_mode` first.. :D

**Comment by lysu** [comment]
LGTM

**Comment by jyz0309** [comment]
@XuHuaiyu PTAL

File: executor/write_test.go [hunk 4]
```diff
@@ -3868,6 +3868,17 @@ func (s *testSerialSuite) TestIssue20840(c *C) {
...
 }
 
+func (s *testSerialSuite) TestIssue22496(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t12")
+	tk.MustExec("set sql_mode=''")
+	tk.MustExec("create table t12(d decimal(15,2));")
+	tk.MustExec("insert into t12 values('1,999.00');")
+	tk.MustQuery("SELECT * FROM t12;").Check(testkit.Rows("0.00"))
```
Comment (XuHuaiyu) [inline executor/write_test.go:3878]: mysql returns 1.00 here?

Comment (jyz0309) [reply executor/write_test.go]: have fixed it，PTAL

File: executor/write_test.go [hunk 5]
```diff
@@ -3868,6 +3868,17 @@ func (s *testSerialSuite) TestIssue20840(c *C) {
 	tk.MustExec("drop table t1")
 }
 
+func (s *testSerialSuite) TestIssue22496(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t12")
+	tk.MustExec("set sql_mode=''")
```
Comment (XuHuaiyu) [inline executor/write_test.go:3875]: add a case for non-empty sql_mode?

File: types/mydecimal.go [hunk 6]
```diff
@@ -389,6 +389,7 @@ func (d *MyDecimal) ToString() (str []byte) {
 
 // FromString parses decimal from string.
 func (d *MyDecimal) FromString(str []byte) error {
+	var strErr error
```
Comment (XuHuaiyu) [inline types/mydecimal.go:392]: Add a comment for this

**Comment by jyz0309** [comment]
@XuHuaiyu Done.PTAL

**Review by XuHuaiyu — APPROVED** [review]
LGTM

**Comment by jyz0309** [comment]
@lysu @XuHuaiyu it seems that the e2e-test's fault is not occured by this pr?PTAL

**Comment by jyz0309** [comment]
@XuHuaiyu PTAL

**Pull request merged**